### PR TITLE
Allow passing a h5py file when opening a file

### LINKED ISSUE #236: Support passing an already open h5py file
This should be possible:
```python
with h5py.File(...) as f:
    with snx.File(f, ...) as nx:
        ...
```

But currently, `snx.File.__init__` simply forwards all arguments to `h5py.File` which does not support this usage pattern.

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #236

### CODE REVIEW

File: src/scippnexus/file.py [hunk 1]
```diff
@@ -27,22 +30,35 @@ def __init__(
 
         Parameters
         ----------
+        name:
+            Path, Bytes object, or h5py.Group.
```
Comment (nvaytet) [inline src/scippnexus/file.py:34]: Can we leave this line out if it's already specified in the type hints? Would avoid it accidentally becoming out of date if we change something in the future...

File: src/scippnexus/file.py [hunk 2]
```diff
@@ -17,6 +19,7 @@
 class File(AbstractContextManager, Group):
     def __init__(
         self,
+        name: str | os.PathLike[str] | io.BytesIO | h5py.Group,
```
Comment (nvaytet) [inline src/scippnexus/file.py:22]: If the file is open, is it a `h5py.Group` or a `h5py.File`?

File: tests/file_test.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+def is_closed(file: snx.File) -> bool:
+    try:
+        file.create_class('x', 'NXentry')
+    except ValueError as err:
+        return 'Unable to synchronously create group' in err.args[0]
+    return False
+
+
+@pytest.mark.parametrize('path_type', [str, Path])
+def test_load_entry_from_filename(tmp_path, path_type):
```
Comment (nvaytet) [inline tests/file_test.py:22]: `tmp_path`: I hate `pytest` and all its hidden conventions/pre-defined objects. It took me a while to figure out where `tmp_path` was defined... :-(

File: tests/file_test.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    with h5.File(buffer, 'w') as f:
+        snx.create_class(f, 'entry', snx.NXentry)
+    buffer.seek(0)
+
+    with snx.File(buffer, 'r') as f:
+        assert f.keys() == {'entry'}
+    assert is_closed(f)
+
+
+def test_load_entry_from_h5py_group_toor(tmp_path):
```
Comment (nvaytet) [inline tests/file_test.py:42]:
```suggestion
def test_load_entry_from_h5py_group_root(tmp_path):
```

File: tests/file_test.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    with h5.File('test.nxs', 'w', driver='core', backing_store=False) as h5_file:
```
Comment (nvaytet) [inline tests/file_test.py:43]: I can't quite figure out if this is creating a file on disk or not. From the `h5py` docs for `backing_store`, it says "If False, any changes are discarded when the file is closed."
Does this mean that if the file did not exist, and you did not make changes to it, then no file will be created, or will an empty file be created?

File: tests/file_test.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        with snx.File(h5_file['entry']) as snx_file:
+            assert snx_file.keys() == {'user'}
+        # Not great, but we don't want to close the h5py file ourselves.
+        assert not is_closed(snx_file)
+
+
+def test_file_from_h5py_group_does_not_allow_extra_args(tmp_path):
+    with h5.File('test.nxs', 'w', driver='core', backing_store=False) as h5_file:
+        snx.create_class(h5_file, 'entry', snx.NXentry)
+        with pytest.raises(TypeError):
```
Comment (nvaytet) [inline tests/file_test.py:64]:
```suggestion
        with pytest.raises(TypeError, match='Cannot provide both h5py.File and other arguments'):
```
?

File: src/scippnexus/file.py (hunk 2 above)
Comment (jl-wynen) [reply src/scippnexus/file.py:22]: `h5py.File`. But that is a subclass of `h5py.group`. So this annotation is enough and also supports passing in a subgroup.

File: tests/file_test.py (hunk 5 above)
Comment (jl-wynen) [reply tests/file_test.py:43]: I don't actually know. This is used everywhere else in the tests. I just copied it from there.

**Review by nvaytet — APPROVED** [review]

**Pull request merged**